### PR TITLE
Set a couple of other parameters needed in createConstitutiveModel

### LINKED ISSUE #13203: Fix handling of parameters in SolidModel::CreateConstitutiveModel()
## Bug Description
In SolidMechanics, there is a design in which materials that derive from SolidModel can create materials that derive from ConstitutiveModel. The way it is currently set up, the two models have to have exactly the same set of parameters. 

There's at least one pair of models that incorrectly puts the parameters for the SolidModel-derived class in the ConstitutiveModel-derived class just to get around this problem, even though some of those parameters are not used by ConstitutiveModel. This means that the user puts those parameters in that block, incorrectly assuming that they do something.

This system is of course going away, and there are some inherent problems with this design, but we need to fix this issue so that we can identify and correct these issues with existing models as we migrate them over to TensorMechanics-based models.

We can fix this by changing `SolidModel::CreateConstitutiveModel()` so that it only passes in the parameters that apply to the ConstitutiveModel class. This will allow us to fix up the sets of parameters in the individual models on a case-by-case basis.

## Steps to Reproduce
See above.

## Impact
The proposed fix to `SolidModel::CreateConstitutiveModel()` won't change any existing model behavior. However, it will allow us to go through and remove parameters that don't apply to classes that derive from ConstitutiveModel. The fact that some of those models have parameters that don't actually do anything is a severe bug because users think they are setting something, but it doesn't have the intended effect. I don't really think there are cases like this in the SolidMechanics module, but if they are, I will address them under this issue.

### PR DESCRIPTION
**Pull request by bwspenc**
closes #13203

### CODE REVIEW

**Comment by bwspenc** [comment]
I'm pretty sure this is the fix we need, but I'm running the assessment case that broke right now, and should get a definitive answer on whether this fixes the problem in a couple of hours.

**Comment by bwspenc** [comment]
@gardnerru FYI I believe this fixes the problems with the assessment cases.

**Review by dschwen — APPROVED** [review]
Looks good to me, provided it fixes the assessments. Aaaand you fix the the missing angle bracket...

**Review by dschwen — CHANGES_REQUESTED** [review]
Missing angle bracket

**Comment by bwspenc** [comment]
@dschwen yeah, the version I had out there didn't actually compile...

The version I pushed passed the assessment problem, so I think it's good to go. I did have to add a check to make sure the parameter was valid, because it's only set for the face and neighbor cases.

File: modules/solid_mechanics/src/materials/SolidModel.C [hunk 1]
```diff
@@ -1596,6 +1596,11 @@ SolidModel::createConstitutiveModel(const std::string & cm_name)
 
   params.applyParameters(parameters());
   params.set<SubProblem *>("_subproblem") = &_subproblem;
+  if (parameters().isParamValid("_material_data_type"))
+    params.set<Moose::MaterialDataType>("_material_data_type") =
+        getParam<Moose::MaterialDataType>("_material_data_type");
+  if (parameters().isParamValid("_neighbor"))
+    params.set<bool>("_neighbor") = getParam<bool>("_neighbor");
```
Comment (permcody) [inline modules/solid_mechanics/src/materials/SolidModel.C]: I don't think I like this fix...  From looking at this tiny piece of code, I'm assuming you are being bitten by rule 4 in the applyParameters() routine (i.e. both (parameters) cannot be private). I can appreciate the logic that @aeslaughter used in building that routine but if it's going to cause issues in practice, perhaps we need to rethink that logic a bit? 

Private parameters often hold things like pointers to members so we probably shouldn't just go and copy them. Maybe one side or the other shouldn't have one of these parameters listed as private? 

Maybe we should have an override for applyParameters that will apply private parameters? I worry that this kind of workaround will bleed further out than we want it to. Any ideas?

Comment (aeslaughter) [reply modules/solid_mechanics/src/materials/SolidModel.C]: How about adding a param specific version of applyParameters:

```C++
params.applyParameters(parameters(), {"_material_data_type", "_neighbor"});
```

Comment (aeslaughter) [reply modules/solid_mechanics/src/materials/SolidModel.C]: I can add this feature if @permcody likes the idea

Comment (dschwen) [reply modules/solid_mechanics/src/materials/SolidModel.C]: Cody, I don't think we should perform any framework fixes to make this more elegant. This pattern (factory calls in a material) is limited to solid mechanics and will soon die along with it. The change that Ben implemented in the previous PR helped clean up some very misleading input files and was intended as a short term fix to make sure current users are getting out of their simulations what they intended. 

I am not in love with this fix, but it is a consequence of this undesirable pattern of creating unregistered material objects. I would strongly advise against accommodating this pattern and would suggest to permit this fix given that it has an expiration date.

Comment (permcody) [reply modules/solid_mechanics/src/materials/SolidModel.C]: @dschwen - OK. Feel free to merge.
@aeslaughter - let's not worry about it then until we have a better use case where there's a compelling reason.

**Review by dschwen — APPROVED** [review]

File: modules/solid_mechanics/src/materials/SolidModel.C (hunk 1 above)
Comment (bwspenc) [reply modules/solid_mechanics/src/materials/SolidModel.C]: Yeah, I'll be the first to admit that this is ugly.  I'm also hoping that we can delete this code really soon!

**Pull request merged**